File: csharp/src/Fory/ReadContext.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

namespace Apache.Fory;

public sealed class ReadContext
{
    private const int MaxParsedTypeMetaEntries = 8192;

    private readonly ReusableArray<TypeMeta> _readTypeMetas = new();
    private readonly Dictionary<ulong, TypeMeta> _cachedTypeMetasByHeader = [];
    private TypeMeta? _firstReadTypeMeta;
    private bool _hasFirstReadTypeMeta;
    private ulong _lastMetaHeader;
    private TypeMeta? _lastTypeMeta;
    private bool _hasLastMetaHeader;

    private readonly List<MetaString> _readMetaStrings = [];

    internal readonly UInt64Map<TypeInfo> _readTypeInfoByType = new();
    internal readonly List<uint> _reservedRefIds = [];
    private readonly int _maxDynamicReadDepth;
    internal Type? _typeMetaType;
    internal TypeMeta? _typeMeta;
    internal UInt64Map<TypeMeta>? _typeMetaByType;
    internal Type? _cachedTypeMetaType;
    internal TypeMeta? _cachedTypeMeta;
    internal int _currentDynamicReadDepth;

    public ReadContext(
        ByteReader reader,
        TypeResolver typeResolver,
        bool trackRef,
        bool compatible = false,
        bool checkStructVersion = false,
        int maxDynamicReadDepth = 20)
    {
        if (maxDynamicReadDepth <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(maxDynamicReadDepth), "MaxDepth must be greater than 0.");
        }

        Reader = reader;
        TypeResolver = typeResolver;
        TrackRef = trackRef;
        Compatible = compatible;
        CheckStructVersion = checkStructVersion;
        RefReader = new RefReader();
        _maxDynamicReadDepth = maxDynamicReadDepth;
    }

    public ByteReader Reader { get; private set; }

    public TypeResolver TypeResolver { get; }

    public bool TrackRef { get; }

    public bool Compatible { get; }

    public bool CheckStructVersion { get; }

    internal RefReader RefReader { get; }

    internal void ResetFor(ByteReader reader)
    {
        Reader = reader;
        Reset();
    }

    internal TypeMeta? GetReadTypeMeta(int index)
    {
        if (index < 0)
        {
            return null;
        }

        if (index == 0)
        {
            return _hasFirstReadTypeMeta ? _firstReadTypeMeta : null;
        }

        return _readTypeMetas.Get(index - 1);
    }

    internal void StoreReadTypeMeta(TypeMeta typeMeta, int index)
    {
        if (index < 0)
        {
            throw new InvalidDataException("negative type meta index");
        }

        if (index == 0)
        {
            _firstReadTypeMeta = typeMeta;
            _hasFirstReadTypeMeta = true;
            return;
        }

        if (!_hasFirstReadTypeMeta)
        {
            throw new InvalidDataException(
                $"type meta index gap: index={index}, missing index 0");
        }

        int listIndex = index - 1;
        if (listIndex == _readTypeMetas.Count)
        {
            _readTypeMetas.Add(typeMeta);
            return;
        }

        if (listIndex < _readTypeMetas.Count)
        {
            _readTypeMetas.Set(listIndex, typeMeta);
            return;
        }

        throw new InvalidDataException(
            $"type meta index gap: index={index}, count={_readTypeMetas.Count + 1}");
    }

    internal bool TryGetCachedReadTypeMeta(ulong header, out TypeMeta typeMeta)
    {
        if (_hasLastMetaHeader && _lastMetaHeader == header && _lastTypeMeta is not null)
        {
            typeMeta = _lastTypeMeta;
            return true;
        }

        if (_cachedTypeMetasByHeader.TryGetValue(header, out TypeMeta? cached) && cached is not null)
        {
            _lastMetaHeader = header;
            _lastTypeMeta = cached;
            _hasLastMetaHeader = true;
            typeMeta = cached;
            return true;
        }

        typeMeta = null!;
        return false;
    }

    internal void CacheReadTypeMeta(ulong header, TypeMeta typeMeta)
    {
        if (_cachedTypeMetasByHeader.TryGetValue(header, out TypeMeta? existing) && existing is not null)
        {
            _lastMetaHeader = header;
            _lastTypeMeta = existing;
            _hasLastMetaHeader = true;
            return;
        }

        if (_cachedTypeMetasByHeader.Count >= MaxParsedTypeMetaEntries)
        {
            return;
        }

        _lastMetaHeader = header;
        _lastTypeMeta = typeMeta;
        _hasLastMetaHeader = true;
        _cachedTypeMetasByHeader.TryAdd(header, typeMeta);
    }

    internal MetaString? GetReadMetaString(int index)
    {
        return index >= 0 && index < _readMetaStrings.Count ? _readMetaStrings[index] : null;
    }

    internal void AppendReadMetaString(MetaString value)
    {
        _readMetaStrings.Add(value);
    }

    internal TypeMeta ReadTypeMeta()
    {
        uint indexMarker = Reader.ReadVarUInt32();
        bool isRef = (indexMarker & 1) == 1;
        int index = checked((int)(indexMarker >> 1));
        if (isRef)
        {
            TypeMeta? cached = GetReadTypeMeta(index);
            if (cached is null)
            {
                throw new InvalidDataException($"unknown type meta ref index {index}");
            }

            return cached;
        }

        ulong header = Reader.ReadUInt64();
        if (TryGetCachedReadTypeMeta(header, out TypeMeta cachedTypeMeta))
        {
            // Header-cache hits intentionally skip without rehashing. Entries reach this cache only
            // after a successful TypeMeta parse and 52-bit body-hash validation. The current body
            // size still comes from the current header bytes, not from the cached TypeMeta.
            TypeMeta.SkipBody(Reader, header);
            StoreReadTypeMeta(cachedTypeMeta, index);
            return cachedTypeMeta;
        }

        Reader.MoveBack(sizeof(ulong));
        TypeMeta typeMeta = TypeMeta.Decode(Reader);
        StoreReadTypeMeta(typeMeta, index);
        CacheReadTypeMeta(header, typeMeta);
        return typeMeta;
    }

    internal void StoreTypeMeta(Type type, TypeMeta typeMeta)
    {
        ulong typeKey = TypeMapKey.Get(type);
        if (_cachedTypeMetaType == type && ReferenceEquals(_cachedTypeMeta, typeMeta))
        {
            return;
        }

        if (ReferenceEquals(_typeMetaType, type))
        {
            if (ReferenceEquals(_typeMeta, typeMeta))
            {
                _cachedTypeMetaType = type;
                _cachedTypeMeta = typeMeta;
                return;
            }

            _typeMeta = typeMeta;
            _cachedTypeMetaType = type;
            _cachedTypeMeta = typeMeta;
            return;
        }

        if (_typeMetaType is null)
        {
            _typeMetaType = type;
            _typeMeta = typeMeta;
            _cachedTypeMetaType = type;
            _cachedTypeMeta = typeMeta;
            return;
        }

        if (_typeMetaByType is null)
        {
            _typeMetaByType = new UInt64Map<TypeMeta>();
            if (_typeMeta is not null)
            {
                _typeMetaByType.Set(TypeMapKey.Get(_typeMetaType!), _typeMeta);
            }
        }
        else if (_typeMetaByType.TryGetValue(typeKey, out TypeMeta? existing) &&
                 ReferenceEquals(existing, typeMeta))
        {
            _cachedTypeMetaType = type;
            _cachedTypeMeta = typeMeta;
            return;
        }

        _typeMetaByType.Set(typeKey, typeMeta);
        _cachedTypeMetaType = type;
        _cachedTypeMeta = typeMeta;
    }

    public TypeMeta? GetTypeMeta<T>()
    {
        return GetTypeMeta(typeof(T));
    }

    private TypeMeta? GetTypeMeta(Type type)
    {
        ulong typeKey = TypeMapKey.Get(type);
        if (_cachedTypeMetaType == type && _cachedTypeMeta is not null)
        {
            return _cachedTypeMeta;
        }

        if (ReferenceEquals(_typeMetaType, type) &&
            _typeMeta is not null)
        {
            _cachedTypeMetaType = type;
            _cachedTypeMeta = _typeMeta;
            return _typeMeta;
        }

        if (_typeMetaByType is null ||
            !_typeMetaByType.TryGetValue(typeKey, out TypeMeta? typeMeta) ||
            typeMeta is null)
        {
            return null;
        }

        _cachedTypeMetaType = type;
        _cachedTypeMeta = typeMeta;
        return typeMeta;
    }

    internal void SetReadTypeInfo(Type type, TypeInfo typeInfo)
    {
        _readTypeInfoByType.Set(TypeMapKey.Get(type), typeInfo);
    }

    internal TypeInfo? GetReadTypeInfo(Type type)
    {
        return _readTypeInfoByType.TryGetValue(TypeMapKey.Get(type), out TypeInfo? typeInfo) ? typeInfo : null;
    }

    internal void ClearReadTypeInfo(Type type)
    {
        _readTypeInfoByType.Remove(TypeMapKey.Get(type));
    }

    public void StoreRef(object? value)
    {
        if (_reservedRefIds.Count == 0)
        {
            return;
        }

        RefReader.StoreRefAt(_reservedRefIds[^1], value);
    }

    internal void SetReservedRefId(uint refId)
    {
        _reservedRefIds.Add(refId);
    }

    internal void ClearReservedRefId()
    {
        if (_reservedRefIds.Count > 0)
        {
            _reservedRefIds.RemoveAt(_reservedRefIds.Count - 1);
        }
    }

    internal void IncreaseReadDepth()
    {
        _currentDynamicReadDepth += 1;
        if (_currentDynamicReadDepth > _maxDynamicReadDepth)
        {
            throw new InvalidDataException(
                $"maximum dynamic object nesting depth ({_maxDynamicReadDepth}) exceeded. current depth: {_currentDynamicReadDepth}");
        }
    }

    internal void DecreaseReadDepth()
    {
        if (_currentDynamicReadDepth > 0)
        {
            _currentDynamicReadDepth -= 1;
        }
    }

    internal void Reset()
    {
        RefReader.Reset();
        _typeMetaType = null;
        _typeMeta = null;
        _typeMetaByType?.ClearKeys();
        _readTypeInfoByType.ClearKeys();
        _reservedRefIds.Clear();
        _cachedTypeMetaType = null;
        _cachedTypeMeta = null;
        _currentDynamicReadDepth = 0;
        _firstReadTypeMeta = null;
        _hasFirstReadTypeMeta = false;
        _readTypeMetas.Clear();
        _readMetaStrings.Clear();
    }
}

File: swift/Sources/Fory/ReadContext.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Foundation

private let typeMetaSizeMask = 0xFF

public final class ReadContext {
    public let buffer: ByteBuffer
    let typeResolver: TypeResolver
    public let xlang: Bool
    public let trackRef: Bool
    public let compatible: Bool
    public let checkClassVersion: Bool
    public let maxCollectionSize: Int
    public let maxBinarySize: Int
    public let maxDepth: Int
    public let refReader: RefReader
    private let compatibleTypeDefTypeInfos = ReusableArray<TypeInfo?>(defaultValue: nil, reserve: 2)
    private let metaStrings = ReusableArray<MetaString?>(defaultValue: nil, reserve: 16)
    private var dynamicAnyDepth = 0

    private var typeInfoStack = UInt64Map<TypeInfo>(initialCapacity: 8)
    private var typeInfoScopeStack: [(typeKey: UInt64, previousTypeInfo: TypeInfo?)] = []
    private var lastTypeInfo = TypeInfo.uncached

    init(
        buffer: ByteBuffer,
        typeResolver: TypeResolver,
        xlang: Bool = false,
        trackRef: Bool,
        compatible: Bool = false,
        checkClassVersion: Bool = true,
        maxCollectionSize: Int = 1_000_000,
        maxBinarySize: Int = 64 * 1024 * 1024,
        maxDepth: Int = 5
    ) {
        self.buffer = buffer
        self.typeResolver = typeResolver
        self.xlang = xlang
        self.trackRef = trackRef
        self.compatible = compatible
        self.checkClassVersion = checkClassVersion
        self.maxCollectionSize = maxCollectionSize
        self.maxBinarySize = maxBinarySize
        self.maxDepth = maxDepth
        self.refReader = RefReader()
    }

    @inline(__always)
    func enterDynamicAnyDepth() throws {
        if maxDepth < 0 {
            throw ForyError.invalidData("configured maxDepth \(maxDepth) is negative")
        }
        let nextDepth = dynamicAnyDepth + 1
        if nextDepth > maxDepth {
            throw ForyError.invalidData(
                "dynamic Any nesting depth \(nextDepth) exceeds configured maxDepth \(maxDepth)"
            )
        }
        dynamicAnyDepth = nextDepth
    }

    @inline(__always)
    func leaveDynamicAnyDepth() {
        if dynamicAnyDepth > 0 {
            dynamicAnyDepth -= 1
        }
    }

    @inline(__always)
    func ensureCollectionLength(_ length: Int, label: String) throws {
        if length < 0 {
            throw ForyError.invalidData("\(label) length is negative")
        }
        if length > maxCollectionSize {
            throw ForyError.invalidData(
                "\(label) length \(length) exceeds configured maxCollectionSize \(maxCollectionSize)"
            )
        }
    }

    @inline(__always)
    func ensureBinaryLength(_ length: Int, label: String) throws {
        if length < 0 {
            throw ForyError.invalidData("\(label) size is negative")
        }
        if length > maxBinarySize {
            throw ForyError.invalidData(
                "\(label) size \(length) exceeds configured maxBinarySize \(maxBinarySize)"
            )
        }
    }

    @inline(__always)
    func ensureRemainingBytes(_ byteCount: Int, label: String) throws {
        if byteCount < 0 {
            throw ForyError.invalidData("\(label) size is negative")
        }
        let remainingBytes = buffer.remaining
        if byteCount > remainingBytes {
            throw ForyError.invalidData(
                "\(label) requires \(byteCount) bytes but only \(remainingBytes) remain in buffer"
            )
        }
    }

    @inline(__always)
    func typeInfo<T: Serializer>(for type: T.Type) throws -> TypeInfo {
        let typeID = ObjectIdentifier(type)
        if lastTypeInfo.swiftTypeID == typeID {
            return lastTypeInfo
        }
        let info = try typeResolver.requireTypeInfo(for: type)
        lastTypeInfo = info
        return info
    }

    @inline(__always)
    func readStaticTypeInfo(_ typeID: TypeId) throws -> TypeInfo? {
        let rawTypeID = UInt32(try buffer.readUInt8())
        guard let actualTypeID = TypeId(rawValue: rawTypeID) else {
            throw ForyError.invalidData("unknown type id \(rawTypeID)")
        }
        if actualTypeID != typeID {
            throw ForyError.typeMismatch(expected: typeID.rawValue, actual: rawTypeID)
        }
        return nil
    }

    func readTypeInfo() throws -> TypeInfo {
        let rawTypeID = UInt32(try buffer.readUInt8())
        guard let wireTypeID = TypeId(rawValue: rawTypeID) else {
            throw ForyError.invalidData("unknown dynamic type id \(rawTypeID)")
        }

        switch wireTypeID {
        case .compatibleStruct, .namedCompatibleStruct:
            return try readCompatibleTypeInfo()
        case .namedEnum, .namedStruct, .namedExt, .namedUnion:
            if compatible {
                return try readCompatibleTypeInfo()
            }
            let namespace = try readMetaString(
                context: self,
                decoder: .namespace,
                encodings: namespaceMetaStringEncodings
            )
            let typeName = try readMetaString(
                context: self,
                decoder: .typeName,
                encodings: typeNameMetaStringEncodings
            )
            return try typeResolver.requireTypeInfo(namespace: namespace.value, typeName: typeName.value)
        case .structType, .enumType, .ext, .typedUnion, .union:
            let userTypeID = try buffer.readVarUInt32()
            return try typeResolver.requireTypeInfo(userTypeID: userTypeID)
        default:
            return typeResolver.builtinTypeInfo(for: wireTypeID)
        }
    }

    func readTypeInfo<T: Serializer>(for type: T.Type) throws -> TypeInfo? {
        let rawTypeID = UInt32(try buffer.readUInt8())
        guard let typeID = TypeId(rawValue: rawTypeID) else {
            throw ForyError.invalidData("unknown type id \(rawTypeID)")
        }

        guard T.staticTypeId.isUserTypeKind else {
            if typeID != T.staticTypeId {
                throw ForyError.typeMismatch(expected: T.staticTypeId.rawValue, actual: rawTypeID)
            }
            return nil
        }

        let localTypeInfo = try typeInfo(for: type)
        let expectedWireTypeID = localTypeInfo.wireTypeID(compatible: compatible)
        if !isAllowedRegisteredWireTypeID(
            typeID,
            declaredTypeID: localTypeInfo.typeID,
            registerByName: localTypeInfo.registerByName,
            compatible: compatible,
            evolving: localTypeInfo.evolving
        ) {
            throw ForyError.typeMismatch(expected: expectedWireTypeID.rawValue, actual: rawTypeID)
        }

        switch typeID {
        case .compatibleStruct, .namedCompatibleStruct:
            return try readCompatibleTypeInfoIfNeeded(
                for: localTypeInfo,
                wireTypeID: typeID
            )
        case .namedEnum, .namedStruct, .namedExt, .namedUnion:
            if compatible {
                _ = try readCompatibleTypeInfoIfNeeded(
                    for: localTypeInfo,
                    wireTypeID: typeID
                )
            } else {
                let namespace = try readMetaString(
                    context: self,
                    decoder: .namespace,
                    encodings: namespaceMetaStringEncodings
                )
                let typeName = try readMetaString(
                    context: self,
                    decoder: .typeName,
                    encodings: typeNameMetaStringEncodings
                )
                guard localTypeInfo.registerByName else {
                    throw ForyError.invalidData("received name-registered type info for id-registered local type")
                }
                if namespace.value != localTypeInfo.namespace.value ||
                    typeName.value != localTypeInfo.typeName.value {
                    let expectedTypeName = "\(localTypeInfo.namespace.value)::\(localTypeInfo.typeName.value)"
                    let actualTypeName = "\(namespace.value)::\(typeName.value)"
                    throw ForyError.invalidData(
                        "type name mismatch: expected \(expectedTypeName), got \(actualTypeName)"
                    )
                }
            }
        default:
            if !localTypeInfo.registerByName && registeredWireTypeNeedsUserTypeID(typeID) {
                guard let localUserTypeID = localTypeInfo.userTypeID else {
                    throw ForyError.invalidData("missing user type id for id-registered type")
                }
                let remoteUserTypeID = try buffer.readVarUInt32()
                if remoteUserTypeID != localUserTypeID {
                    throw ForyError.typeMismatch(expected: localUserTypeID, actual: remoteUserTypeID)
                }
            }
        }
        return nil
    }

    @inline(__always)
    private func readCompatibleTypeInfoIfNeeded(
        for localTypeInfo: TypeInfo,
        wireTypeID: TypeId
    ) throws -> TypeInfo? {
        let buffer = self.buffer
        let compatibleTypeDefTypeInfos = self.compatibleTypeDefTypeInfos
        if !checkClassVersion,
           compatibleTypeDefTypeInfos.isEmpty,
           !localTypeInfo.typeDefHasUserTypeFields,
           let localTypeDefHeader = localTypeInfo.typeDefHeader {
            let typeMetaStart = buffer.getCursor()
            let indexMarker = try buffer.readVarUInt32()
            if indexMarker == 0 {
                let header = try buffer.readUInt64()
                var bodySize = Int(header & UInt64(typeMetaSizeMask))
                if bodySize == typeMetaSizeMask {
                    bodySize += Int(try buffer.readVarUInt32())
                }
                if header == localTypeDefHeader {
                    // Header-cache hits intentionally skip without rehashing. Entries reach this
                    // cache only after a successful TypeDef parse and 52-bit metadata-hash validation.
                    compatibleTypeDefTypeInfos.push(localTypeInfo)
                    try buffer.skip(bodySize)
                    return nil
                }
            }
            buffer.setCursor(typeMetaStart)
        }
        return try readCompatibleTypeInfo(
            for: localTypeInfo,
            wireTypeID: wireTypeID
        )
    }

    private func readCompatibleTypeInfo() throws -> TypeInfo {
        let buffer = self.buffer
        let compatibleTypeDefTypeInfos = self.compatibleTypeDefTypeInfos
        let indexMarker = try buffer.readVarUInt32()
        let isRef = (indexMarker & 1) == 1
        let index = Int(indexMarker >> 1)
        if isRef {
            guard let typeInfo = compatibleTypeDefTypeInfos.get(index) else {
                throw ForyError.invalidData("unknown compatible type definition ref index \(index)")
            }
            return typeInfo
        }

        let typeMetaStart = buffer.getCursor()
        let header = try buffer.readUInt64()
        var bodySize = Int(header & UInt64(typeMetaSizeMask))
        if bodySize == typeMetaSizeMask {
            bodySize += Int(try buffer.readVarUInt32())
        }
        if let cached = typeResolver.getTypeInfo(forHeader: header) {
            // Header-cache hits intentionally skip without rehashing. Entries reach this cache only
            // after a successful TypeDef parse and 52-bit metadata-hash validation.
            try buffer.skip(bodySize)
            compatibleTypeDefTypeInfos.push(cached)
            return cached
        }

        buffer.setCursor(typeMetaStart)
        let decoded = try TypeMeta.decode(buffer)
        let cachedTypeInfo = try typeResolver.cacheTypeInfo(decoded, forHeader: header)
        compatibleTypeDefTypeInfos.push(cachedTypeInfo)
        return cachedTypeInfo
    }

    @inline(__always)
    private func readCompatibleTypeInfo(
        for localTypeInfo: TypeInfo,
        wireTypeID: TypeId
    ) throws -> TypeInfo {
        let buffer = self.buffer
        let compatibleTypeDefTypeInfos = self.compatibleTypeDefTypeInfos
        let remoteTypeInfo: TypeInfo
        if compatibleTypeDefTypeInfos.isEmpty,
           let localTypeDefHeader = localTypeInfo.typeDefHeader {
            let typeMetaStart = buffer.getCursor()
            let indexMarker = try buffer.readVarUInt32()
            if indexMarker != 0 {
                buffer.setCursor(typeMetaStart)
                remoteTypeInfo = try readCompatibleTypeInfo()
            } else {
                let header = try buffer.readUInt64()
                var bodySize = Int(header & UInt64(typeMetaSizeMask))
                if bodySize == typeMetaSizeMask {
                    bodySize += Int(try buffer.readVarUInt32())
                }

                if header == localTypeDefHeader {
                    // Header-cache hits intentionally skip without rehashing. Entries reach this
                    // cache only after a successful TypeDef parse and 52-bit metadata-hash validation.
                    compatibleTypeDefTypeInfos.push(localTypeInfo)
                    try buffer.skip(bodySize)
                    return localTypeInfo
                }

                buffer.setCursor(typeMetaStart)
                remoteTypeInfo = try readCompatibleTypeInfo()
            }
        } else {
            remoteTypeInfo = try readCompatibleTypeInfo()
        }
        guard let remoteTypeMeta = remoteTypeInfo.compatibleTypeMeta else {
            throw ForyError.invalidData("compatible type metadata is required")
        }
        if let localTypeMeta = localTypeInfo.typeMeta,
           remoteTypeMeta === localTypeMeta {
            return localTypeInfo
        }
        if remoteTypeMeta.registerByName {
            guard localTypeInfo.registerByName else {
                throw ForyError.invalidData("received name-registered compatible metadata for id-registered local type")
            }
            if remoteTypeMeta.namespace.value != localTypeInfo.namespace.value {
                throw ForyError.invalidData(
                    "namespace mismatch: expected \(localTypeInfo.namespace.value), got \(remoteTypeMeta.namespace.value)"
                )
            }
            if remoteTypeMeta.typeName.value != localTypeInfo.typeName.value {
                throw ForyError.invalidData(
                    "type name mismatch: expected \(localTypeInfo.typeName.value), got \(remoteTypeMeta.typeName.value)"
                )
            }
        } else {
            guard !localTypeInfo.registerByName else {
                throw ForyError.invalidData("received id-registered compatible metadata for name-registered local type")
            }
            guard let remoteUserTypeID = remoteTypeMeta.userTypeID else {
                throw ForyError.invalidData("missing user type id in compatible type metadata")
            }
            guard let localUserTypeID = localTypeInfo.userTypeID else {
                throw ForyError.invalidData("missing local user type id metadata for id-registered type")
            }
            if remoteUserTypeID != localUserTypeID {
                throw ForyError.typeMismatch(expected: localUserTypeID, actual: remoteUserTypeID)
            }
        }

        if let remoteTypeID = remoteTypeMeta.typeID,
           let remoteWireTypeID = TypeId(rawValue: remoteTypeID),
           !isAllowedRegisteredWireTypeID(
               remoteWireTypeID,
               declaredTypeID: localTypeInfo.typeID,
               registerByName: localTypeInfo.registerByName,
               compatible: compatible,
               evolving: localTypeInfo.evolving
           ) {
            throw ForyError.typeMismatch(expected: wireTypeID.rawValue, actual: remoteTypeID)
        }
        return remoteTypeInfo
    }

    func readAnyValue(typeInfo: TypeInfo) throws -> Any {
        try enterDynamicAnyDepth()
        defer { leaveDynamicAnyDepth() }

        let value: Any
        switch typeInfo.typeID {
        case .bool:
            value = try Bool.foryRead(self, refMode: .none, readTypeInfo: false)
        case .int8:
            value = try Int8.foryRead(self, refMode: .none, readTypeInfo: false)
        case .int16:
            value = try Int16.foryRead(self, refMode: .none, readTypeInfo: false)
        case .int32:
            value = try buffer.readInt32()
        case .varint32:
            value = try Int32.foryRead(self, refMode: .none, readTypeInfo: false)
        case .int64:
            value = try buffer.readInt64()
        case .varint64:
            value = try Int64.foryRead(self, refMode: .none, readTypeInfo: false)
        case .taggedInt64:
            value = try buffer.readTaggedInt64()
        case .uint8:
            value = try UInt8.foryRead(self, refMode: .none, readTypeInfo: false)
        case .uint16:
            value = try UInt16.foryRead(self, refMode: .none, readTypeInfo: false)
        case .uint32:
            value = try buffer.readUInt32()
        case .varUInt32:
            value = try UInt32.foryRead(self, refMode: .none, readTypeInfo: false)
        case .uint64:
            value = try buffer.readUInt64()
        case .varUInt64:
            value = try UInt64.foryRead(self, refMode: .none, readTypeInfo: false)
        case .taggedUInt64:
            value = try buffer.readTaggedUInt64()
        case .float16:
            value = try Float16.foryRead(self, refMode: .none, readTypeInfo: false)
        case .bfloat16:
            value = try BFloat16.foryRead(self, refMode: .none, readTypeInfo: false)
        case .float32:
            value = try Float.foryRead(self, refMode: .none, readTypeInfo: false)
        case .float64:
            value = try Double.foryRead(self, refMode: .none, readTypeInfo: false)
        case .string:
            value = try String.foryRead(self, refMode: .none, readTypeInfo: false)
        case .duration:
            value = try Duration.foryRead(self, refMode: .none, readTypeInfo: false)
        case .timestamp:
            value = try Date.foryRead(self, refMode: .none, readTypeInfo: false)
        case .date:
            value = try LocalDate.foryRead(self, refMode: .none, readTypeInfo: false)
        case .decimal:
            value = try Decimal.foryRead(self, refMode: .none, readTypeInfo: false)
        case .binary:
            value = try Data.foryRead(self, refMode: .none, readTypeInfo: false)
        case .boolArray:
            value = try readPrimitiveArray(self) as [Bool]
        case .int8Array:
            value = try readPrimitiveArray(self) as [Int8]
        case .int16Array:
            value = try readPrimitiveArray(self) as [Int16]
        case .int32Array:
            value = try readPrimitiveArray(self) as [Int32]
        case .int64Array:
            value = try readPrimitiveArray(self) as [Int64]
        case .uint8Array:
            value = try readPrimitiveArray(self) as [UInt8]
        case .uint16Array:
            value = try readPrimitiveArray(self) as [UInt16]
        case .uint32Array:
            value = try readPrimitiveArray(self) as [UInt32]
        case .uint64Array:
            value = try readPrimitiveArray(self) as [UInt64]
        case .float16Array:
            value = try readPrimitiveArray(self) as [Float16]
        case .bfloat16Array:
            value = try readPrimitiveArray(self) as [BFloat16]
        case .float32Array:
            value = try readPrimitiveArray(self) as [Float]
        case .float64Array:
            value = try readPrimitiveArray(self) as [Double]
        case .array, .list:
            value = try readListOfAny(refMode: .none) ?? []
        case .set:
            value = try Set<AnyHashable>.foryRead(self, refMode: .none, readTypeInfo: false)
        case .map:
            value = try readDynamicAnyMapValue(context: self)
        case .none:
            value = ForyAnyNullValue()
        default:
            if typeInfo.typeID.isUserTypeKind {
                value = try typeInfo.read(self)
            } else {
                throw ForyError.invalidData("unsupported dynamic type id \(typeInfo.typeID)")
            }
        }
        return value
    }

    @inline(__always)
    func getTypeInfo<T: Serializer>(for type: T.Type) -> TypeInfo? {
        typeInfoStack.value(for: UInt64(UInt(bitPattern: ObjectIdentifier(type))))
    }

    func withTypeInfo<T: Serializer, R>(
        _ typeInfo: TypeInfo?,
        for type: T.Type,
        _ body: () throws -> R
    ) rethrows -> R {
        guard let typeInfo else {
            return try body()
        }

        let typeKey = UInt64(UInt(bitPattern: ObjectIdentifier(type)))
        let previousTypeInfo = typeInfoStack.value(for: typeKey)
        typeInfoScopeStack.append((typeKey: typeKey, previousTypeInfo: previousTypeInfo))
        typeInfoStack.set(typeInfo, for: typeKey)
        defer {
            if let scope = typeInfoScopeStack.popLast() {
                if let previousTypeInfo = scope.previousTypeInfo {
                    typeInfoStack.set(previousTypeInfo, for: scope.typeKey)
                } else {
                    _ = typeInfoStack.removeValue(for: scope.typeKey)
                }
            } else {
                assertionFailure("type info scope stack underflow")
            }
        }
        return try body()
    }

    @inline(__always)
    func getReadMetaString(at index: Int) -> MetaString? {
        metaStrings.get(index)
    }

    @inline(__always)
    func appendReadMetaString(_ value: MetaString) {
        metaStrings.push(value)
    }

    func reset() {
        if dynamicAnyDepth != 0 {
            dynamicAnyDepth = 0
        }
        if trackRef {
            refReader.reset()
        }
        if !typeInfoStack.isEmpty {
            typeInfoStack.clear()
        }
        if !typeInfoScopeStack.isEmpty {
            typeInfoScopeStack.removeAll(keepingCapacity: true)
        }
        compatibleTypeDefTypeInfos.reset()
        metaStrings.reset()
    }
}

public extension ReadContext {
    func readAny(
        refMode: RefMode,
        readTypeInfo: Bool = true
    ) throws -> Any? {
        try SerializableAny.foryRead(self, refMode: refMode, readTypeInfo: readTypeInfo).anyValue()
    }

    func readListOfAny(
        refMode: RefMode,
        readTypeInfo: Bool = false
    ) throws -> [Any]? {
        let wrapped: [SerializableAny]? = try [SerializableAny]?.foryRead(
            self,
            refMode: refMode,
            readTypeInfo: readTypeInfo
        )
        return wrapped?.map { $0.anyValueForCollection() }
    }

    func readMapStringToAny(
        refMode: RefMode,
        readTypeInfo: Bool = false
    ) throws -> [String: Any]? {
        let wrapped: [String: SerializableAny]? = try [String: SerializableAny]?.foryRead(
            self,
            refMode: refMode,
            readTypeInfo: readTypeInfo
        )
        guard let wrapped else {
            return nil
        }
        var map: [String: Any] = [:]
        map.reserveCapacity(wrapped.count)
        for pair in wrapped {
            map[pair.key] = pair.value.anyValueForCollection()
        }
        return map
    }

    func readMapInt32ToAny(
        refMode: RefMode,
        readTypeInfo: Bool = false
    ) throws -> [Int32: Any]? {
        let wrapped: [Int32: SerializableAny]? = try [Int32: SerializableAny]?.foryRead(
            self,
            refMode: refMode,
            readTypeInfo: readTypeInfo
        )
        guard let wrapped else {
            return nil
        }
        var map: [Int32: Any] = [:]
        map.reserveCapacity(wrapped.count)
        for pair in wrapped {
            map[pair.key] = pair.value.anyValueForCollection()
        }
        return map
    }

    func readMapAnyHashableToAny(
        refMode: RefMode,
        readTypeInfo: Bool = false
    ) throws -> [AnyHashable: Any]? {
        let wrapped: [AnyHashable: SerializableAny]? = try [AnyHashable: SerializableAny]?.foryRead(
            self,
            refMode: refMode,
            readTypeInfo: readTypeInfo
        )
        guard let wrapped else {
            return nil
        }
        var map: [AnyHashable: Any] = [:]
        map.reserveCapacity(wrapped.count)
        for pair in wrapped {
            map[pair.key] = pair.value.anyValueForCollection()
        }
        return map
    }
}
